File: java/src/main/java/net/snowflake/floe/BaseSegmentProcessor.java
package net.snowflake.floe;

import java.io.Closeable;
import java.util.Arrays;
import java.util.function.Supplier;

abstract class BaseSegmentProcessor implements Closeable {
  protected static final int NON_TERMINAL_SEGMENT_SIZE_MARKER = -1;
  protected static final int headerTagLength = 32;

  protected final FloeParameterSpec parameterSpec;
  protected final FloeIv floeIv;
  protected final MessageKey messageKey;
  protected final FloeAad floeAad;

  protected final KeyDerivator keyDerivator;

  private AeadKey currentAeadKey;

  private boolean isClosed;
  private boolean completedExceptionally;

  BaseSegmentProcessor(FloeParameterSpec parameterSpec, FloeIv floeIv, FloeKey floeKey, FloeAad floeAad) {
    // Check the key length if we can
    if (floeKey.getKey().getFormat().equalsIgnoreCase("RAW")) {
      byte[] rawKey = null;
      try {
        rawKey = floeKey.getKey().getEncoded();
        // We don't care about contents, only the length
        Arrays.fill(rawKey, (byte) 0);
      } catch (final Exception e) {
        // If we cannot access the key then we must trust it is correct
        rawKey = null;
      }
      if (rawKey != null && rawKey.length != parameterSpec.getAead().getKeyLength()) {
        throw new IllegalArgumentException("invalid key length");
      }
    }
    this.parameterSpec = parameterSpec;
    this.floeIv = floeIv;
    this.floeAad = floeAad;
    this.keyDerivator = new KeyDerivator(parameterSpec);
    this.messageKey = keyDerivator.hkdfExpandMessageKey(floeKey, floeIv, floeAad);
  }

  protected byte[] processInternal(Supplier<byte[]> processFunc) {
    assertNotClosed();
    try {
      byte[] result = processFunc.get();
      completedExceptionally = false;
      return result;
    } catch (FloeException e) {
      completedExceptionally = true;
      throw e;
    } catch (Exception e) {
      completedExceptionally = true;
      throw new FloeException(e);
    }
  }

  protected AeadKey getKey(MessageKey messageKey, FloeIv floeIv, FloeAad floeAad, long segmentCounter) {
    if (currentAeadKey == null || segmentCounter % parameterSpec.getKeyRotationMask() == 0) {
      // we don't need masking, because we derive a new key only when key rotation happens
      currentAeadKey = deriveKey(messageKey, floeIv, floeAad, segmentCounter);
    }
    return currentAeadKey;
  }

  private AeadKey deriveKey(MessageKey secretKey, FloeIv floeIv, FloeAad floeAad, long segmentCounter) {
    return keyDerivator.hkdfExpandAeadKey(
            secretKey,
            floeIv,
            floeAad,
            segmentCounter
    );
  }

  protected void closeInternal() {
    isClosed = true;
  }

  private void assertNotClosed() {
    if (isClosed) {
      throw new IllegalStateException("stream has already been closed");
    }
  }

  @Override
  public void close() {
    if (!isClosed && !completedExceptionally) {
      throw new IllegalStateException("last segment was not processed");
    }
  }

  protected boolean isClosed() {
    return isClosed;
  }
}

File: cpp/src/BaseSegmentProcessor.cpp
#include "BaseSegmentProcessor.hpp"
#include "FloeKey.hpp"
#include "FloeIv.hpp"
#include "FloeAad.hpp"
#include "AeadKey.hpp"
#include "KeyDerivator.hpp"
#include "floe/FloeException.hpp"

// Required as the IDE does not properly understand that the MessageKey is used in the implementation
// ReSharper disable once CppUnusedIncludeDirective
#include "MessageKey.hpp"

namespace floe {

BaseSegmentProcessor::~BaseSegmentProcessor() = default;

BaseSegmentProcessor::BaseSegmentProcessor(
    const FloeParameterSpec& parameterSpec,
    const FloeIv& floeIv,
    const FloeKey& floeKey,
    const FloeAad& floeAad)
    : parameterSpec_(parameterSpec),
      isClosed_(false),
      completedExceptionally_(false) {
    
    if (floeKey.getKeyData().size() != static_cast<size_t>(parameterSpec.getAead().getKeyLength())) {
        throw FloeException("invalid key length");
    }
    
    floeIv_ = std::make_unique<FloeIv>(floeIv.getBytes());
    floeAad_ = std::make_unique<FloeAad>(floeAad.getBytes().data(), floeAad.getBytes().size());
    keyDerivator_ = std::make_unique<KeyDerivator>(parameterSpec);
    messageKey_ = keyDerivator_->hkdfExpandMessageKey(floeKey, floeIv, floeAad);
}

std::vector<uint8_t> BaseSegmentProcessor::processInternal(
    const std::function<std::vector<uint8_t>()> &processFunc) {
    
    assertNotClosed();
    try {
        std::vector<uint8_t> result = processFunc();
        completedExceptionally_ = false;
        return result;
    } catch (const FloeException&) {
        completedExceptionally_ = true;
        throw;
    } catch (const std::exception& e) {
        completedExceptionally_ = true;
        throw FloeException(e);
    }
}

std::unique_ptr<AeadKey> BaseSegmentProcessor::getKey(
    const MessageKey& messageKey,
    const FloeIv& floeIv,
    const FloeAad& floeAad,
    const uint64_t segmentCounter) {
    
    if (!currentAeadKey_ || segmentCounter % parameterSpec_.getKeyRotationMask() == 0) {
        currentAeadKey_ = deriveKey(messageKey, floeIv, floeAad, segmentCounter);
    }
    return std::make_unique<AeadKey>(currentAeadKey_->getKeyData(), currentAeadKey_->getAlgorithm());
}

std::unique_ptr<AeadKey> BaseSegmentProcessor::deriveKey(
    const MessageKey& messageKey,
    const FloeIv& floeIv,
    const FloeAad& floeAad,
    const uint64_t segmentCounter) const {
    
    return keyDerivator_->hkdfExpandAeadKey(messageKey, floeIv, floeAad, segmentCounter);
}

void BaseSegmentProcessor::closeInternal() {
    isClosed_ = true;
}

void BaseSegmentProcessor::assertNotClosed() const {
    if (isClosed_) {
        throw FloeException("stream has already been closed");
    }
}

void BaseSegmentProcessor::close() {
    if (!isClosed_ && !completedExceptionally_) {
        throw FloeException("last segment was not processed");
    }
}

}